### ISSUE TITLE
AzureFileCopyV2 failed when uploading to Azure Storage Static Website

### ISSUE THREAD

**Issue by alaconda** [Microsoft/vsts-tasks#7611]
Hi, 

With the new release of [Azure Static Website](https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/storage/blobs/storage-blob-static-website.md).  I can't seem to find a way to upload to its default container of $web.  Same setting would work if I switch to a different container.

`
Upload to container: '$web' in storage account: '<account name>' with blob prefix: '' failed with error: 'AzCopy.exe exited with non-zero exit code while uploading files to blob storage.' For more info please refer to https://aka.ms/azurefilecopyreadme
`

Is there something extra I need to do?  Or is it not supported yet?  And if so, is it in the pipeline to be released soon?

Thanks!

**Comment by onpaws** [comment]
HI @rajatagrawal-dev I'm curious.
For VSTS users who are interested in testing out the new release of AzureFileCopyV2, is there a way to opt-in before public release? I'd be happy to help test.
I've already opted into the other 'preview' VSTS features but curious about VSTS Tasks specifically.
Thanks

**Comment by rajatagrawal-dev** [comment]
@onpaws AzureFileCopyV2 (using AzCopy v7.1.0) is already publicly available. When you add Azure File Copy task in your build/release definition, you should see an option of selecting the version of the task. You can select version 2 to test the preview version.

**Comment by alaconda** [comment]
Hi @rajatagrawal-dev,

Thanks for the link.  Please keep us posted and let us know if there is anything else we can do to help.

**Comment by rajatagrawal-dev** [comment]
Sure. I'll update when we've made progress on this.

**Comment by nathanvv** [comment]
The file copy with AzureFileCopyV2 doesnt seem to copy the file into the "$web" folder . Logs : 

2018-07-04T16:35:34.3835676Z Uploading files from source path: 'C:\agent\_work\46\s\allure-publish\migration\prod\2018.7.4.22\index.html' to storage account: 'somerandomname' in container: '$web' with blob prefix: ''
2018-07-04T16:35:34.3990429Z ##[command] & "AzCopy\AzCopy.exe" /Source:"C:\agent\_work\46\s\allure-publish\migration\prod\2018.7.4.22" /Dest:"https://somerandomname.blob.core.windows.net/$web" /DestKey:"*****" /XO /Y /SetContentType /Z:"AzCopy" /V:"AzCopy\AzCopyVerbose_03702a45-1109-4fdd-9066-5ee504653e34.log" /S /Pattern:"index.html"
2018-07-04T16:35:34.5146935Z [2018/07/04 16:35:34][WARNING] The command line ""C:\agent\_work\_tasks\AzureFileCopy_eb72cb01-a7e5-427b-a8a1-1b31ccac8a43\2.0.1\AzCopy\AzCopy.exe" /Source:C:\agent\_work\46\s\allure-publish\migration\prod\2018.7.4.21 /Dest:https://somerandomname.blob.core.windows.net//migration /DestKey:****** /XO /Y /SetContentType /Z:AzCopy /V:AzCopy\AzCopyVerbose_a7e93134-e33d-4219-a3ff-a502ec4e6d51.log /S /Pattern:index.html" in the journal file "AzCopy\AzCopy.jnl" is different from your input.
2018-07-04T16:35:34.5147946Z [2018/07/04 16:35:34][WARNING] Incomplete operation with different command line detected at the journal directory "AzCopy".


notice how the " /Dest:https://somerandomname.blob.core.windows.net//migration" is missing the $web

**Comment by nathanvv** [comment]
I have the storage account configured as static website, just running into the issue of $web missing in the url and hence the files being copied into $root. Appreciate the help. Thanks a lot again.

**Comment by onpaws** [comment]
Hi @rajatagrawal-dev,
Thanks for posting back on how to select the 2.x FileCopy task. 
I'm actually already using that with no luck on the $web as other are reporting. 
Was offering to help test out the next release after 2.x (3.x or whatever it's called?)
Either way thanks for your efforts, I know at least a couple people at my company who will be excited to get this

**Comment by rajatagrawal-dev** [comment]
@onpaws Support for $web for static websites will be available a few after AzCopy tool is updated. It will be a minor version update only, so it will also be version 2.x. I'll let you know the exact version of task you should see when we have made the changes.

**Comment by sonu2alok** [comment]
I tried with 2.x FileCopy task but still getting the same error:

Upload to container: '$web' in storage account: 'azaueastdevXXXXX' with blob prefix: '' failed with error: 'AzCopy.exe exited with non-zero exit code while uploading files to blob storage.' For more info please refer to https://aka.ms/azurefilecopyreadme

**Comment by SumiranAgg** [comment]
@sonu2alok The task currently doesn't support static website. We will update the thread once the enhancement is in place.

**Comment by kenotron** [comment]
Looks like the azcopy 7.30 just came out. Can you please update this task is support that version?

**Comment by Jaxwood** [comment]
@rajatagrawal-dev how can I tell which version my account is running?

**Comment by rajatagrawal-dev** [comment]
@Jaxwood One way is to run the task. If you see the release logs of your task, you'll find the version of the task near the top of the logs.
![image](https://user-images.githubusercontent.com/21691254/42632266-4172b0f4-85fa-11e8-885d-21d22e8f0a6f.png)

Another way is to get data from tasks api. Just enter the following in the URL (assuming you are logged in to your account):
https://{you_account_name}.visualstudio.com/_apis/Distributedtask/tasks. This will give you a json of task versions in your account. Search for "AzureFileCopy" (with quotes) in that json and you should find 2 search results (for V1 and V2 of the task). Version will be there right after the search match locations.

**Comment by nathanvv** [comment]
Just curious, looks like we are still at v2.0.1. Is there a release log where I could check to get the date of release of v2.0.1 ?. More curious to get a rough idea oh how long it would take for us to get this patch

![image](https://user-images.githubusercontent.com/11080029/42640908-f386f328-85b8-11e8-8577-a0b5ee47df60.png)

**Comment by rajatagrawal-dev** [comment]
@nathanvv Patch versions of tasks are updated on almost every check-in of code. So the patch versions can easily jump multiple versions between 2 releases. The latest release hasn't been rolled out to all accounts so you might see v2.0.4 in a few days. 
For v2.0.6 and above, it will take at least 2-3 weeks more for all accounts to get updated.

**Comment by denis1stomin** [comment]
Great job!
I've just checked new AzCopy today and it works as expected with $web container.
Will wait till new version will be available for my account. Thanks!

**Comment by alaconda** [comment]
@rajatagrawal-dev, thanks for the update.  Looking forward to > v2.0.6  once it rolled out to my account.

**Comment by markusdresch** [comment]
now it does transfer files from the root directory to $root (instead of $web) and then it fails with:

```
2018-07-25T09:59:33.5323085Z Uploading files from source path: 'D:\a\1\s\src\xxx\wwwroot' to storage account: 'xxx' in container: '$web' with blob prefix: ''
2018-07-25T09:59:33.5545963Z ##[command] & "AzCopy\AzCopy.exe" /Source:"D:\a\1\s\src\xxx\wwwroot" /Dest:"https://xxx.blob.core.windows.net/$web" /@:"D:\a\_temp\xxx" /XO /Y /SetContentType /Z:"AzCopy" /V:"AzCopy\AzCopyVerbose_xxx.log" /S
2018-07-25T09:59:39.2893962Z [2018/07/25 09:59:39][ERROR] Error parsing destination location "https://xxx.blob.core.windows.net/": The remote server returned an error: (400) Bad Request. For more details, please type "AzCopy /?:Dest" or use verbose option /V.
2018-07-25T09:59:39.2916948Z [2018/07/25 09:59:39][ERROR] HttpStatusMessage:The requested URI does not represent any resource on the server.
2018-07-25T09:59:39.2917359Z RequestId:xxx
2018-07-25T09:59:39.2919288Z Time:Wed, 25 Jul 2018 09:59:37 GMT
2018-07-25T09:59:39.2934672Z [2018/07/25 09:59:39] Transfer summary:
2018-07-25T09:59:39.2934932Z -----------------
2018-07-25T09:59:39.2935843Z Total files transferred: 2
2018-07-25T09:59:39.2936898Z Transfer successfully:   2
2018-07-25T09:59:39.2937949Z Transfer skipped:        0
2018-07-25T09:59:39.2938749Z Transfer failed:         0
2018-07-25T09:59:39.2939543Z Elapsed time:            00.00:00:05
2018-07-25T09:59:40.7827898Z ##[error]Upload to container: '$web' in storage account: 'xxx' with blob prefix: '' failed with error: 'AzCopy.exe exited with non-zero exit code while uploading files to blob storage.' For more info please refer to https://aka.ms/azurefilecopyreadme
```
subdirectories are omitted.

**Comment by rajatagrawal-dev** [comment]
@markusdresch In what version of the task are you seeing this behavior? You'll find the task version near the top of the logs.

**Comment by markusdresch** [comment]
```
2018-07-25T09:58:15.8633365Z Task         : Azure File Copy
2018-07-25T09:58:15.8633592Z Description  : Copy files to Azure blob or VM(s)
2018-07-25T09:58:15.8633884Z Version      : 2.0.4
```

**Comment by rajatagrawal-dev** [comment]
Deployment is a bit delayed. It is planned to complete by next week. I'll post more updates when I have any.

**Comment by markusdresch** [comment]
works now with v2.0.7

**Comment by rajatagrawal-dev** [comment]
Static website support should be available in all accounts now. Please check if you are seeing Azure File Copy v2.0.7 in your builds/releases and whether you are able to deploy to $web container successfully.

**Comment by jonathantower** [comment]
It's getting close to working, but what's that backtick doing in there?

![image](https://user-images.githubusercontent.com/1853927/44188869-103e9e80-a0ee-11e8-862f-f885fc08f562.png)

**Comment by johnhydemtm365** [comment]
Hi still not working, unless I have mis-configured something... :-(

**Version is 2.0.7**
2018-08-16T08:18:28.4657458Z ##[section]Starting: AzureBlob File Copy
2018-08-16T08:18:28.4665256Z 
2018-08-16T08:18:28.4665631Z Task         : Azure File Copy
2018-08-16T08:18:28.4665961Z Description  : Copy files to Azure blob or VM(s)
2018-08-16T08:18:28.4666259Z Version      : 2.0.7
2018-08-16T08:18:28.4666564Z Author       : Microsoft Corporation
2018-08-16T08:18:28.4666921Z Help         : [More Information](https://aka.ms/azurefilecopyreadme)
2018-08-16T08:18:28.4667301Z 

**Error:**
2018-08-16T08:20:02.4436158Z ##[command] & "AzCopy\AzCopy.exe" /Source:"D:\a\1\s\public" /Dest:"https://*******************.blob.core.windows.net/`$web" /@:"D:\a\_temp\stnrkcbu06m5c9msjptfyldi" /S
2018-08-16T08:20:04.4304443Z [2018/08/16 08:20:04][ERROR] Error parsing destination location "https://****************.blob.core.windows.net/$web": Failed to validate destination. The remote server returned an error: (403) Forbidden.
2018-08-16T08:20:04.4305539Z HttpStatusMessage:This request is not authorized to perform this operation.
2018-08-16T08:20:04.4306049Z RequestId:b6273607-101e-0017-5139-355a77000000
2018-08-16T08:20:04.4306550Z Time:Thu, 16 Aug 2018 08:20:03 GMT For more details, please type "AzCopy /?:Dest" or use verbose option /V.


Thanks
John

**Comment by rajatagrawal-dev** [comment]
@johnhydemtm365 Does this only happen with $web container or with any container name? Can you re-execute with some other container name like "testContainer" instead of "$web"?

Also, please enable debug logging. Add a variable in your release definition called 'system.debug' and set it's value to 'true'. Then create a new release to get debug logs.

**Comment by johnhydemtm365** [comment]
Hi nope, does not work with a normal container. Updated AzCopy Locally and it all works fine in a windows batch file.
Log File:
[Vsts_AzcopyTaskFailure.txt](https://github.com/Microsoft/vsts-tasks/files/2293687/Vsts_AzcopyTaskFailure.txt)

**Comment by rajatagrawal-dev** [comment]
@johnhydemtm365 Since the issue does not seem to be related to static website support but rather related to permissions, can you open a new issue for this to get the right traction. Please attach the logs in that as well.

**Comment by jonathantower** [comment]
@rajatagrawal-dev I figured out my issue, and it is definitely something you'll want to fix. In the "Azure File COpy (Preview)" task, I had mistakenly specified the Source to be "/", thinking it was a relative path. I noticed in the logs that it was copying from c:\ on the underlying VM! When I changed Source to $(Build.SourcesDirectory), it worked, but I thought you'd want to know about this issue.

**Comment by rajatagrawal-dev** [comment]
@jonathantower We expect you to give absolute paths in source path input or use pre-defined build/release variables like $(Build.SourcesDirectory) and $(System.ArtifactsDirectory). This is also specified in the help text of the input.
"/" and "\\" paths in Windows OS usually correspond to the root C:\ path only . If you execute `cd /` from a command prompt in Windows it takes you to "C:\\". So this behaviour is expected by design.

**Comment by ociata** [comment]
Sorry for posting in this thread, however I've got the very same issue as original issuer. What is frustrating me, is how hard is to debug the problem.

Attached is my full error log, however the interesting part is:
`2018-10-10T07:05:57.4949064Z [2018/10/10 07:05:57][ERROR] Error parsing source location "D:\a\r1\a\Internal-Blog": Failed to enumerate directory D:\a\r1\a\Internal-Blog\ with file pattern *. The system cannot find the path specified. (Exception from HRESULT: 0x80070003) For more details, please type "AzCopy /?:Source" or use verbose option /V.
`
I guess it is related to my setup of Source:
[http://prntscr.com/l4brys](url)

Yet I'm completely lost of how to explore and/or download my build artefact to see how to properly setup this path. Other option will be to somehow list/browse/explore the machine to see what is on  `D:\a\r1\a\I`. I've found following link describing how to do just that:
[https://docs.microsoft.com/en-us/azure/devops/pipelines/build/artifacts?view=vsts](url)

... yet my UI is completely different and I cannot found aforementioned options:
[http://prntscr.com/l4btiw](url)

Any help will be highly appreciated, and if this is not the proper thread to ask please advise how to proceed.

Regards,

**Comment by ociata** [comment]
Okay, I figured it out - yet I wish the documentation was a little bit more helpful.

So I was missing publish artefacts from my azure-pipeline yaml:
```
pool:
  vmImage: 'Ubuntu 16.04'

steps:
- task: NodeTool@0
  inputs:
    versionSpec: '8.x'
  displayName: 'Install Node.js'

- script: |
    cd srv
    npm install
    npm run build
  displayName: 'npm install and build'

# This was missing
- task: PublishBuildArtifacts@1
  inputs:
    PathtoPublish: '$(System.DefaultWorkingDirectory)/srv/public'
```
After I added this step, my UI now properly displays the published artefact and I can download and/or explore it:
[http://prntscr.com/l4pemh](http://prntscr.com/l4pemh)

One more important thing I found the hard way - you can only deploy to AzureBlob from Windows Hosted machine (I had mine setup with Ubuntu). The throwing error is something like

> you have script build for Windows only such as powershell

I haven't, but obviously the uploading job has :) 

Hope this is useful to someone in the future. 

Regards,

**Comment by omeshp** [comment]
@ociata Which task are you using to deploy to AzureBlob which is giving the powershell error?

**Comment by ociata** [comment]
> @ociata Which task are you using to deploy to AzureBlob which is giving the powershell error?

AzureBlob File Copy (Preview)

**Comment by rajatagrawal-dev** [comment]
@ociata You can try using the Azure CLI task to upload to blob on linux agents.
Please refer [this](https://docs.microsoft.com/en-us/cli/azure/storage/blob?view=azure-cli-latest#az-storage-blob-upload) for details on using Azure CLI to upload to blob.

**Comment by timtlm** [comment]
I am using this feature to deploy my static website using Azure DevOps, and I have it working properly with v2, but I want to be able to delete everything before deploying.  Currently, if I remove a file from my project, it will still remain in the folder after this pipeline task.  
There is some mention of being able to do this here: https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/deploy/azure-file-copy
Release pipelines don't support yaml, so it isn't very helpful.  However, I tried placing /cleanTargetBeforeCopy:true in the optional arguments and it says the option is not recognized.

**Comment by rajatagrawal-dev** [comment]
@timtlm The cleanTargetBeforeCopy option is for when you are copying to a VM. It is not for copying to blob storage.
Right now AzCopy tool (which Azure FIle Copy task uses internally) doesn't seem to provide the option to delete the container. A workaround will be to add a Azure Powershell task before the Azure File Copy task in your release pipeline and execute the following command:
`Remove-AzureRmStorageContainer -ResourceGroupName "myResourceGroup" -AccountName "myStorageAccount" -ContainerName "myContainer"`
Refer [this](https://docs.microsoft.com/en-us/powershell/module/azurerm.storage/remove-azurermstoragecontainer?view=azurermps-6.10.0).
This will delete the entire container. The container will be automatically recreated when Azure File Copy task is executed.

**Comment by timtlm** [comment]
@rajatagrawal-dev Thanks for the follow-up and the info.  I ended up using the Azure CLI task with the following command:
`az storage blob delete-batch --account-name <storage_account_name> --source $web`
It worked as needed.  I felt a little uncomfortable removing the $web container as it seems to have some special properties that go along with it, but I'm guessing your solution would work as well.

**Comment by chandanch** [comment]
This issue is occuring in version 3.* and 4.*(preview):

![image](https://user-images.githubusercontent.com/8663552/78333037-bc284c80-75a6-11ea-80b3-5efa24c908b9.png)


![image](https://user-images.githubusercontent.com/8663552/78333010-af0b5d80-75a6-11ea-92bd-00535802c9bd.png)

**Comment by ChristosMonogios** [comment]
> This issue is occuring in version 3.* and 4.*(preview):
> 
> ![image](https://user-images.githubusercontent.com/8663552/78333037-bc284c80-75a6-11ea-80b3-5efa24c908b9.png)
> 
> ![image](https://user-images.githubusercontent.com/8663552/78333010-af0b5d80-75a6-11ea-92bd-00535802c9bd.png)

Hey, I managed to get this running and deploying my static app to a Storage Account. Version used was 4.*.

My first problem was that I had no publish-step in my build for my website, [check this solution](https://developercommunity.visualstudio.com/content/problem/337714/can-build-code-but-release-pipeline-says-no-artifa.html)

The second problem was a 403 authentication error from Azure. You need to set the Storage Blob Data Contributor & the Storage Blob Data Owner rights for your App, [check this from another issue](https://github.com/Azure/azure-storage-azcopy/issues/120)
Check also your Service Connections in your Azure DevOps Project Settings.

**Comment by Lienfa** [comment]
I got the same error today

just downgrade task version to 2.* can solve this issue

**Comment by danielmariz** [comment]
> I got the same error today
> 
> just downgrade task version to 2.* can solve this issue

Me too, I've also tried 4 (preview) version that does not work. Only version 2

**Comment by adrien-constant** [comment]
I also got this error, and I got it working by setting the **Storage Blob Data Contributor** role to the connection principal of the service connection, as @ChristosMonogios mentioned.

**Comment by PedroC88** [comment]
I have the same issue, I tried versions but the first and I get the same error in all of them, this definetly shouldn't be closed.

**Comment by JQUINONES82** [comment]
Had the same issue..Reverting back to version 2 works.  Version 3 and 4 do not work for me.

**Comment by tstevanovic** [comment]
Me too, version 2 works. Version 3 and 4 does not work.

**Comment by PontiusTheBarbarian** [comment]
Had a 403, validated everything as detailed by @ChristosMonogios and still failed. Dropped the version to 2* and worked first time.

**Comment by mwanchap** [comment]
Assigning the "Storage Blob Data Contributor" role to the ARM service connection's service principal as @adrien-constant so helpfully suggested worked, using version 4.171.3.

My YAML:
```yaml
- task: AzureFileCopy@4
  displayName: 'Copy files to blob storage: $(storageName)'
  inputs:
    SourcePath: '$(build.sourcesDirectory)\archive\'
    Destination: AzureBlob
    storage: $(storageName)
    ContainerName: $web
    azureSubscription: 'ARM service connection goes here' # needs a role assignment before it'll work
```

**Comment by securityvoid** [comment]
@mwanchap What do you mean by:
`# needs a role assignment before it'll work`

What do I need to assign where for it to work?

**Comment by mwanchap** [comment]
@securityvoid the service principal associated with the ARM connection you're using has to have access to the storage resource in Azure RBAC. The easiest way to get to the service principal is to navigate to your Azure Devops project settings, select "Service Connections", select the service connection you're using and then click "Manage Service Principal". That'll show you its name and ID in Azure AD. Then you'll need to go to the storage resource you want to grant it access to, and give the service principal the "Storage Blob Data Contributor" role on that resource.

**Comment by chrisg32** [comment]
Ran into this problem today. Assigning the "Storage Blob Data Contributor" role did not help. Dropped to version three seemed to do the trick.

```yaml
- task: AzureFileCopy@3
  inputs:
    SourcePath: 'HelpDocumentation'
    azureSubscription: 'Subscription Name'
    Destination: 'AzureBlob'
    storage: 'storageaccountname'
    ContainerName: '$web'
```

**Comment by JFMarty** [comment]
Can confirm in 2023, only Version 3.x seems to work. Couldn't get it to run with Versions 4.x and 5.x, even with additional roles as described above.

Maybe it's because the service principal expired (secret) and I had to re-create the connection?
